### PR TITLE
Provide gzipped release tarball as well (#132)

### LINKED ISSUE #132: Provide gzipped release tarball as well
We should also provide gzipped tarballs for those who don't have BZip2 built into tar(1).

### PR DESCRIPTION
**Pull request by michael-o**
Split the dist process by producing an uncompressed tarball and then compress
with gzip(1) and bzip2(1). Also remove those tarballs on 'make distclean'.

This closes #132

### CODE REVIEW

File: Makefile.in [hunk 1]
```diff
@@ -62,7 +62,12 @@ install: all
 distcheck:
 	@$(ECHO) @PACKAGE_VERSION@
 
-dist: config.h.in configure
-	tar -s +^+msktutil-@PACKAGE_VERSION@/+ -jcvf msktutil-@PACKAGE_VERSION@.tar.bz2 \
+tar: config.h.in configure
+	tar -s +^+msktutil-@PACKAGE_VERSION@/+ -cvf msktutil-@PACKAGE_VERSION@.tar \
```
Comment (dkobras) [inline Makefile.in:66]: If we go this route and keep the tar archive as an intermediate file, we can avoid the phony target and just use the filename directly, ie.

```
@PACKAGE_NAME@-@PACKAGE_VERSION@.tar: config.h.in configure
	tar -s +^+@PACKAGE_NAME@-@PACKAGE_VERSION@/+ -cvf $@ $(SOURCE) $(HEADERS) (...)
```

Unrelated to the actual change, but I wonder whether we can improve on hardcoding the `-s` option which is specific to BSD `tar`. (With GNU `tar` we could use `tar --transform s+^+@PACKAGE_NAME@-@PACKAGE_VERSION@/+ (...)` instead.) I'll see to come up with an additional cleanup.

File: Makefile.in [hunk 2]
```diff
@@ -62,7 +62,12 @@ install: all
...
                 $(SOURCES) $(HEADERS) configure README LICENSE ChangeLog INSTALL \
                 Makefile.in configure.ac config.h.in msktutil.M autogen.sh
+
+dist: tar
+	gzip -kf msktutil-@PACKAGE_VERSION@.tar
+	bzip2 -zkf msktutil-@PACKAGE_VERSION@.tar
```
Comment (dkobras) [inline Makefile.in:72]: Likewise, I'd rather see these in a more canonic form like eg.:

```
%.gz: %
       gzip -c $< > $@

%.bz2: %
       bzip2 -c $< > $@

dist: @PACKAGE_NAME@-@PACKAGE_VERSION@.tar.gz @PACKAGE_NAME@-@PACKAGE_VERSION@.tar.bz2
```

**Comment by dkobras** [comment]
@michael-o I've updated the PR based on the comments above. `make dist` now creates an archive with the "best" available compression format by default, `make dist-all` generates all available formats, and `make dist-xz dist-bzip2 dist-gzip` can be used to force specific formats (if available), which might be handy when automating the release process from a CI pipeline.

I've also switched the generation of the tarball from BSD `tar` to GNU `tar` syntax, while preferring `gnutar` and `gtar` over plain `tar`. The limited set of BSD-compatible machines (that also serve as build hosts) I'm familiar can cope this way. Please let me know if this poses a problem in your environment, meaning we'll have to implement a better workaround.

File: Makefile.in [hunk 3]
```diff
@@ -40,6 +48,27 @@ $(PROG): $(OBJECTS)
...
+else
+	$(BZIP2) -c $< > $@
+endif
+
+%.xz: %
+ifeq ($(XZ),)
+	$(error No xz compression available)
+else
+	$(XZ) -c $< > $@
+endif
```
Comment (michael-o) [inline Makefile.in:70]: Well, with that this Makefile is not POSIX anymore. We have explicitly removed GNU extensions with Olaf back then to make it POSIX only.

File: Makefile.in [hunk 4]
```diff
@@ -20,14 +20,22 @@ SED=@SED@
...
+BZIP2=@BZIP2@
+XZ=@XZ@
 MAN=$(PROG).1
 SOURCES_CPP=krb5wrap.cpp msktutil.cpp msktkrb5.cpp msktldap.cpp msktname.cpp msktpass.cpp msktconf.cpp ldapconnection.cpp
 SOURCES_C=strtoll.c
 SOURCES=$(SOURCES_CPP) $(SOURCES_C)
 HEADERS=msktutil.h msktname.h krb5wrap.h ldapconnection.h
 OBJECTS=$(SOURCES_CPP:.cpp=.o) $(SOURCES_C:.c=.o)
+DISTTARGETS=$(addprefix dist-,$(and $(XZ),xz) $(and $(BZIP2),bzip2) $(and $(GZIP),gzip))
+DISTTARGET=$(firstword $(DISTTARGETS))
```
Comment (michael-o) [inline Makefile.in:34]: Same here for portability

**Comment by michael-o** [comment]
Here is the POSIX fix: https://github.com/msktutil/msktutil/commit/fb308658b8a887124b90e93bebd94f461b1d95f8#diff-b48650cabc0f502b52865f525765f2d688198f06e0ffb5456c76fabab7f2bfb2

File: Makefile.in (hunk 3 above)
Comment (dkobras) [reply Makefile.in]: The latest changeset converts those to shell conditionals, and adds an explicit `.POSIX:` statement. So syntax seems be fine, and current build rules should be unaffected. Only the `dist` targets will require a `make` implementation with suitable extensions, but then building the source distribution is architecture independent by definition.

**Comment by dkobras** [comment]
Thinking some more about this, given there's no easy one-size-fits-all solution because of the `-s` vs. `--transform` incompatibility between BSD and GNU `tar`, I'd favour a simpler solution checking for `gtar`/`gnutar` if available and just using native compression as before. I'll push a revised changeset on top of yours. Please let me know what you think.

**Review by dkobras — APPROVED** [review]

**Comment by michael-o** [comment]
Why not run sed in a separate step?

File: Makefile.in [hunk 5]
```diff
@@ -48,7 +51,9 @@ clean:
 	$(RM) $(PROG) $(OBJECTS) $(MAN)
 
 distclean: clean
-	$(RM) Makefile config.h config.log config.cache config.status autom4te.cache config.h~ config.h.in~ config.h.in
+	$(RM) Makefile config.h config.log config.cache config.status autom4te.cache \
+	      config.h~ config.h.in~ config.h.in \
+	      $(DISTARCHIVE).tar.gz $(DISTARCHIVE).tar.bz2
```
Comment (michael-o) [inline Makefile.in:56]: Will rm fail on nob existing files? Maybe -f?

Comment (michael-o) [reply Makefile.in:56]: not*

Comment (dkobras) [reply Makefile.in:56]: `autoconf` initializes `RM` as `/bin/rm -rf`, so this is already taken care of.

**Comment by dkobras** [comment]
> Why not run sed in a separate step?

We supply `tar` with the actual path to the file, but then add a prefix (subdir) to the path _within the archive itself_. This is internal to `tar`, and cannot be processed with an external tool. If we were to drop `--transform` or `-s`, we'd have to do something like (untested and probably full of GNUisms as well):

```
# create adequately named tmpdir structure
$(MKDIR) -p $(SOME_TMPDIR)/$(DISTARCHIVE)
# copy content to tmpdir
$(TAR) -c $^ $(EXTRA_DIST) | $(TAR) -x -C $(SOME_TMPDIR)/$(DISTARCHIVE)
# create archive (from within tmpdir) and store resulting file in current dir
$(TAR) -jcv -C $(SOME_TMPDIR) $(DISTARCHIVE) > $@
# clean up
$(RM) $(SOME_TMPDIR)
```

File: Makefile.in [hunk 6]
```diff
@@ -62,7 +67,10 @@ install: all
...
 
-dist: config.h.in configure
-	tar -s +^+msktutil-@PACKAGE_VERSION@/+ -jcvf msktutil-@PACKAGE_VERSION@.tar.bz2 \
-                $(SOURCES) $(HEADERS) configure README LICENSE ChangeLog INSTALL \
-                Makefile.in configure.ac config.h.in msktutil.M autogen.sh
+$(DISTARCHIVE).tar.gz: $(SOURCES) $(HEADERS) config.h.in configure
+	$(TAR) --transform=s+^+$(DISTARCHIVE)/+ -jcvf $@ $^ $(EXTRA_DIST)
+
+$(DISTARCHIVE).tar.bz2: $(SOURCES) $(HEADERS) config.h.in configure
+	$(TAR) --transform=s+^+$(DISTARCHIVE)/+ -zcvf $@ $^ $(EXTRA_DIST)
```
Comment (dkobras) [inline Makefile.in:74]: Oops, obviously `-j` and `-z` should be the other way around. I'll fix up.

**Comment by michael-o** [comment]
@dkobras Ran locally with changes to BSD tar:
```
$ tar tfz msktutil-1.1.tar.gz
msktutil-1.1/README
msktutil-1.1/LICENSE
msktutil-1.1/ChangeLog
msktutil-1.1/INSTALL
msktutil-1.1/Makefile.in
msktutil-1.1/configure.ac
msktutil-1.1/msktutil.M
msktutil-1.1/autogen.sh
```
sources and headers are gone from the tarball: $(SOURCES) $(HEADERS)

**Comment by dkobras** [comment]
@michael-o Argh, so it turns out that `$^` is not portable, either. Fixed.

**Comment by michael-o** [comment]
Squashed and merged. @dkobras Thank you!

**Pull request closed without merging**